pkg/storkctl/action.go

```diff
@@ -20,99 +21,150 @@ const (
+	return performCommands
+}
+
+var mockTime *time.Time
+
+// setMockTime is used in tests to update the time
+func setMockTime(mt *time.Time) {
+	mockTime = mt
+}
+
+// GetCurrentTime returns the current time as per the scheduler
+func GetCurrentTime() time.Time {
+	if mockTime != nil {
+		return *mockTime
+	}
+	return time.Now()
 }
 
 func newFailoverCommand(cmdFactory Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
-	failoverCommand := &cobra.Command{
+	var referenceMigrationSchedule string
+	var skipDeactivateSource bool
+	var includeNamespaceList []string
+	var excludeNamespaceList []string
+	var namespaceList []string
+	performFailoverCommand := &cobra.Command{
 		Use:   failoverCommand,
-		Short: "Initiate failover for the given namespaces",
+		Short: "Initiate failover of the given migration schedule",
 		Run: func(c *cobra.Command, args []string) {
-			namespaces, err := cmdFactory.GetAllNamespaces()
+			if len(referenceMigrationSchedule) == 0 {
+				util.CheckErr(fmt.Errorf("referenceMigrationSchedule name needs to be provided for failover"))
+				return
+			}
+			// namespace of the migrationSchedule is provided by user using the -n / --namespace global flag
+			migrationScheduleNs := cmdFactory.GetNamespace()
+			migrSchedObj, err := storkops.Instance().GetMigrationSchedule(referenceMigrationSchedule, migrationScheduleNs)
 			if err != nil {
-				util.CheckErr(err)
+				util.CheckErr(fmt.Errorf("unable to find the referenceMigrationSchedule %v in the %v namespace", referenceMigrationSchedule, migrationScheduleNs))
 				return
 			}
-			failedToStart := false
-			for _, namespace := range namespaces {
-				if incompleteAction := getAnyIncompleteAction(namespace); incompleteAction != nil {
-					if !failedToStart {
-						printMsg("Failed to start failover as there pending actions for following namespaces:", ioStreams.Out)
-						failedToStart = true
-					}
-					printMsg(fmt.Sprintf(
-						"Namespace %v has action %v in state %v",
-						namespace, incompleteAction.Name, incompleteAction.Status),
-						ioStreams.Out)
+			if !skipDeactivateSource {
+				clusterPair := migrSchedObj.Spec.Template.Spec.ClusterPair
+				_, err = storkops.Instance().GetClusterPair(clusterPair, migrationScheduleNs)
+				if err != nil {
+					util.CheckErr(fmt.Errorf("unable to find the cluster pair %v in the %v namespace", clusterPair, migrationScheduleNs))
+					return
 				}
 			}
-			if failedToStart {
+			migrationNamespaceList := migrSchedObj.Spec.Template.Spec.Namespaces
+			migrationNamespaceSelectors := migrSchedObj.Spec.Template.Spec.NamespaceSelectors
+			// update the migrationNamespaces list by fetching namespaces based on provided label selectors
+			migrationNamespaces, err := getMigrationNamespaces(migrationNamespaceList, migrationNamespaceSelectors)
+			if err != nil {
+				util.CheckErr(fmt.Errorf("unable to get the namespaces based on the --namespace-selectors in the provided migrationSchedule: %v", err))
```

Comment:
The helper function being called here, `getMigrationNamespaces` returns the migrationNamespaceList itself in case selectors is empty. So no possibility of an error in that scenario.
The function's responsibility is to append the namespaces based on the selectors to the provided list of namespaces. So errors can only arise if we are unable to get the namespaces based on the selectors.
```
namespaces, err := core.Instance().ListNamespaces(map[string]string{key: val})
```
 That's why the error message conveys the same.